teps/0044-data-locality-and-pod-overhead-in-pipelines.md

```diff
@@ -145,59 +145,170 @@ See [TEP-0074](./0074-deprecate-pipelineresources.md) for the deprecation plan f
+  * [When expressions](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#guard-task-execution-using-whenexpressions)
+    (and [Conditions](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#guard-task-execution-using-conditions))
+  * [Finally tasks](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#adding-finally-to-the-pipeline)
+  * [Allowing step failure](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-onerror-for-a-step)
+* Functionality that would require significantly expanded orchestration logic:
+  * [Custom tasks](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#using-custom-tasks) - the pod would
+    need to be able to create and watch Custom Tasks, or somehow lean on the Pipelines controller to do this
+* Functionality that might not be possible (i.e. constrained by pods themselves):
+  * Any dynamically created TaskRuns. This is discussed in more detail [below](#dynamically-created-tasks-in-pipelines).
+  * Running each Task with a different `ServiceAccount` - the pod has one ServiceAccount as a whole
+
+(See also [functionality supported by experimental Pipeline to TaskRun](https://github.com/tektoncd/experimental/tree/main/pipeline-to-taskrun#supported-pipeline-features))
+
+### Dynamically created Tasks in Pipelines
+Currently, the number of Tasks in a Pipeline is determined at authoring time based on the number of items in a Pipeline's `Tasks` field.
+However, [TEP-0090: Matrix](./0090-matrix.md) proposes a feature that would allow additional Tasks to be created at runtime.
+In summary, a Pipeline may need to run a Task multiple times with different parameters. The parameters "fanned out" to a Task run
+multiple times in parallel may be specified at authoring time, or they may come from an earlier Task. For example, a Pipeline might clone a repo,
+read a set of parameters from the repo, and run a build or test task once for each of these parameters.
+
+A controller responsible for running multiple Tasks in one pod must know how many Tasks will be run before creating the pod.
+This is because a pod will start executing once it has been created, and many of the fields (including the containers list)
+[cannot be updated](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core).
+However, the number of Tasks needed may not be known until the previous Task is run and its outputs are retrieved.
+Therefore, we may not be able to support running a Pipeline with a matrix of parameters in one pod, or other forms of dynamic Task creation.
+
+### Additional Design Considerations
+- Executing an entire Pipeline in a pod, as compared to executing multiple Tasks in a pod, may pave the way for supporting
+[local execution](https://github.com/tektoncd/pipeline/issues/235).
+
+## Design proposal
 
 TBD - currently focusing on enumerating and examining alternatives before selecting one or more ways forward.
 
 ## Alternatives
```

Comment:
NIT: I wonder if it would help building some kind of matrix of the various options.
Not all pros/cons/supported functionality would fit in the matrix, but it might help us visualise the data?
Not something for this PR in any case, maybe something we can build collaboratively or so.

Also it might make sense to identify if all options are orthogonal or if any of the options could make sense to live together.

Reply:
good idea, I can put together some sort of doc, but agreed that it's not for this PR.